AnimationNodeAnimation use speed_scale set in AnimationPlayer

guzzard opened the pull request:
This change makes animation nodes take the playback speed set in the animation player into account so that animation nodes in state machines can be played slower and faster by changing the speed in the animation player.

Not sure if this is the correct approach design wise, but I needed a way to slow down/speed up animations that played in a state machine and this was an easy way to achieve that.

Please give me feedback, and let me know if this is something you might want to merge, and if I should make any changes, thanks.

guilhermefelipecgs commented:
You need to squash your commits.

https://docs.godotengine.org/en/3.1/community/contributing/pr_workflow.html#mastering-the-pr-workflow-the-rebase

guzzard committed:
AnimationNodeAnimation use speed_scale set in AnimationPlayer

This change makes animation nodes take the playback speed set in the animation player into account so what animations in state machines can be played slower and faster by changing the speed in the animation player
Cleanup


Merge remote-tracking branch 'guzzard-godot/animationnodeanimation-speed-scale' into animationnodeanimation-speed-scale

slapin commented:
Why don't you use BlendTree to control speed of your animations?
This PR adds unnecessary complication and inconsistency. Have you tested that it does not break normal way of things (i.e setting speed via BlendTree's TimeScale node)?

guzzard commented:
@slapin I have not found any smart way of updating the time scale in the time scale node from script during run-time. Since the time scale node inherits from Object (non built-in type) it's not possible to attach a script to the node which checks the an object and updates the time scale based on a value etc? How would you go about solving this?

The reason I chose the solution in the PR is because the animtion tree holds a reference to the animation player, and the animation tree takes the animation players state every update, so it seemed like a somewhat logical path..

slapin commented:
I mean you can implement this directly using AnimationTree nodes. Why you
add second conflicting way to change time scale?

On Mon, Mar 18, 2019 at 1:26 PM Martin W <notifications@github.com> wrote:

> @slapin <https://github.com/slapin> I have not found any smart way of
> updating the time scale in the time scale node from script during run-time.
> Since the time scale node inherits from Object (non built-in type) it's not
> possible to attach a script to the node which checks the an object and
> updates the time scale based on a value etc? How would you go about solving
> this?
>
> The reason I chose the solution in the PR is because the animtion tree
> holds a reference to the animation player, and the animation tree takes the
> animation players state every update, so it seemed like a somewhat logical
> path..
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/godotengine/godot/pull/27049#issuecomment-473852878>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAX05mPbPM8JM8E2INjEh0z9P6zw85vks5vX2nfgaJpZM4b0R3K>
> .
>

slapin commented:
if your root is BlendTree "BT", and your StateMachine name is SM, changing
data from script will be:
$StateMachine["parameters/BT/SM/scale_node/amount"] = 0.1

or something like that (writing from memory). To no exactly the parameter
name, go to your AnimationTree (click it)
that will give you inspector window, there look for "Parameters" open that
and find node and value there.
Hover your mouse over value name and you see complete path displayed. This
way you can change all node parameters in tree.
Not new user friendly but very efficient.

On Mon, Mar 18, 2019 at 1:46 PM Sergey Lapin <slapinid@gmail.com> wrote:

> I mean you can implement this directly using AnimationTree nodes. Why you
> add second conflicting way to change time scale?
>
> On Mon, Mar 18, 2019 at 1:26 PM Martin W <notifications@github.com> wrote:
>
>> @slapin <https://github.com/slapin> I have not found any smart way of
>> updating the time scale in the time scale node from script during run-time.
>> Since the time scale node inherits from Object (non built-in type) it's not
>> possible to attach a script to the node which checks the an object and
>> updates the time scale based on a value etc? How would you go about solving
>> this?
>>
>> The reason I chose the solution in the PR is because the animtion tree
>> holds a reference to the animation player, and the animation tree takes the
>> animation players state every update, so it seemed like a somewhat logical
>> path..
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/godotengine/godot/pull/27049#issuecomment-473852878>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AAAX05mPbPM8JM8E2INjEh0z9P6zw85vks5vX2nfgaJpZM4b0R3K>
>> .
>>
>

guilhermefelipecgs commented:
@slapin is right, it's already possible to control the time scale of state machine using a blend tree like this:
![Captura de tela de 2019-03-18 09-29-39](https://user-images.githubusercontent.com/1387165/54530055-73343480-4960-11e9-934f-b81f7db451c9.png)

guilhermefelipecgs commented:
@guzzard And you can change the time scale for this example by code this way `$animation_tree['parameters/TimeScale/scale'] = 2` or if the script Inherits from `AnimationTree`, then you can use this `set('parameters/TimeScale/scale', 2)`

guzzard commented:
@slapin @guilhermefelipecgs Thanks for your kind help :smiley: 

Did some testing but cant get it to work though.. :disappointed: 

```stateMachinePlayback.set("parameters/Move/TimeScale/scale", 10.0)``` - no error but speed doesn't change
```stateMachinePlayback["parameters/Move/TimeScale/scale"] = 10.0``` - error: Invalid set index 'parameters/Move/TimeScale/scale' (on base: 'AnimationNodeStateMachinePlayback') with value of type 'float'.

guzzard commented:
@slapin @guilhermefelipecgs also trying to read the current value returns null? ```var current_value = stateMachinePlayback.get("parameters/Move/TimeScale/scale")```

slapin commented:
If you can wait till tomorrow I will come up with step-by-step example.
Until then please submit bug report to documentation as it is very sparse.
If enough people report bugs on AT docs some people might stop being jerks
and the thing finally get documentation it deserves.

On Mon, Mar 18, 2019 at 3:56 PM Martin W <notifications@github.com> wrote:

> @slapin <https://github.com/slapin> @guilhermefelipecgs
> <https://github.com/guilhermefelipecgs> also trying to read the current
> value returns null? var current_value =
> stateMachinePlayback.get("parameters/Move/TimeScale/scale")
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/godotengine/godot/pull/27049#issuecomment-473898502>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAX00nFRiwoCEFL930BNT7aoRAN8JLBks5vX40JgaJpZM4b0R3K>
> .
>

guilhermefelipecgs commented:
@guzzard Here is an example: [time_scale.zip](https://github.com/godotengine/godot/files/2978475/time_scale.zip)

Press space to change the time scale.

guzzard commented:
@guilhermefelipecgs awesome, thanks. Understand what my error was now. Tried to set the parameters in AnimationNodeStateMachinePlayback when they need to be set in AnimationTree. Not very user friendly system currently when you need to deal with so many different nodes to achieve what you want, instead of having just one clean interface if you understand what I mean. At least now I understand how to work with it, thanks! Closing this PR.

slapin commented:
Please do not forget to submit bug to documentation.

On Mon, Mar 18, 2019 at 7:32 PM Martin W <notifications@github.com> wrote:
>
> Closed #27049.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

guzzard commented:
@slapin not sure how to phrase that bug report though since there was no bug? Perhaps it's better if you open up an issue since you have a better idea of how to phrase it? I would be happy to chime in and support in comments.

slapin commented:
read AnimationTree doc and explain in the report what was not clear to you.
The doc has a lot to be desired, should not be a problem for you to find
things.

The doc:
https://docs.godotengine.org/en/3.1/tutorials/animation/animation_tree.html

On Tue, Mar 19, 2019 at 12:01 AM Martin W <notifications@github.com> wrote:

> @slapin <https://github.com/slapin> not sure how to phrase that bug
> report though since there was no bug? Perhaps it's better if you open up an
> issue since you have a better idea of how to phrase it? I would be happy to
> chime in and support in comments.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/godotengine/godot/pull/27049#issuecomment-474099435>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAX0zJHg5dH3JcLpgAkQuHIjD2jFUciks5vX_7DgaJpZM4b0R3K>
> .
>

Gyeff commented:
This was very helpful. I encountered a similar issue and fixed it using blend tree.
